File: samples/python/tutorial_code/imgcodecs/exif_test.py
import cv2 as cv
import numpy as np

def dump_exif(exif_entries, title="EXIF DUMP"):
    """
    Print all EXIF entries in a readable format.
    """
    print(f"=== {title} ===")
    for block in exif_entries:
        print('----------------------')
        for entry in block:
            print(entry.dumpAsString())

def get_exif_block(metadata_types, metadata):
    """
    Find the EXIF metadata block from metadata.
    """
    for i, t in enumerate(metadata_types):
        if t == cv.IMAGE_METADATA_EXIF:
            return metadata[i][0]
    return None

def print_exif(path):
    # Load image and metadata
    img, metadata_types, metadata = cv.imreadWithMetadata(path)
    if img is None:
        print("Failed to load image:", path)
        return

    # Extract the raw EXIF block
    raw_exif = get_exif_block(metadata_types, metadata)
    if raw_exif is None:
        print("No EXIF metadata found")
        return

    # Decode EXIF into structured entries
    result, exif_entries = cv.decodeExif(raw_exif)
    if not result:
        print("Parsing EXIF raw data failed")
        return

    # Dump original
    dump_exif(exif_entries, title="ORIGINAL EXIF")

    # Encode with original metadata
    _, buf = cv.imencodeWithMetadata(".jpg", img, metadata_types, metadata)

    # Re-encode EXIF
    res, raw_exif_data = cv.encodeExif(exif_entries)
    if not res:
        print("Failed to encode EXIF")
        return

    new_metadata = [np.frombuffer(raw_exif_data, dtype=np.uint8), metadata[1], metadata[2]]

    _, buf = cv.imencodeWithMetadata(".jpg", img, metadata_types, new_metadata)

    # Reload and verify
    _, metadata_types2, metadata2 = cv.imdecodeWithMetadata(buf)
    raw_exif2 = get_exif_block(metadata_types2, metadata2)

    _, exif_entries2 = cv.decodeExif(raw_exif2)
    print("*" * 60)
    dump_exif(exif_entries2, title="RELOADED EXIF")

if __name__ == "__main__":
    print_exif(cv.samples.findFile("board.jpg"))

File: samples/cpp/tutorial_code/imgcodecs/exif_test.cpp
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include <iostream>

using namespace cv;
using namespace std;

// Function to print EXIF entries in a readable format
static void dumpExif(const std::vector<std::vector<ExifEntry>>& exif_entries, const std::string& title = "EXIF DUMP")
{
    cout << "=== " << title << " ===" << endl;
    for (const auto& block : exif_entries)
    {
        cout << "----------------------" << endl;
        // Dump each entry using ExifEntry's dump method
        for (const auto& entry : block)
        {
            std::cout << entry.dumpAsString() << std::endl;
        }
    }
}

// Function to extract EXIF metadata block from metadata
static std::vector<uchar> getExifBlock(const std::vector<int>& metadata_types, const std::vector<std::vector<uchar>>& metadata)
{
    for (size_t i = 0; i < metadata_types.size(); i++)
    {
        if (metadata_types[i] == IMAGE_METADATA_EXIF)
        {
            return metadata[i];  // Return EXIF buffer
        }
    }
    return {};  // Return empty vector if no EXIF found
}

// Function to read, print, re-encode, and verify EXIF metadata
static void printExif(const std::string& path)
{
    cv::Mat img;
    std::vector<int> metadata_types;          // Holds metadata type codes
    std::vector<std::vector<uchar>> metadata; // Holds raw metadata buffers

    // Load image with metadata
    img = cv::imreadWithMetadata(path, metadata_types, metadata);
    if (img.empty())
    {
        cout << "Failed to load image: " << path << endl;
        return;
    }

    // Extract raw EXIF data
    std::vector<uchar> raw_exif = getExifBlock(metadata_types, metadata);
    if (raw_exif.empty())
    {
        cout << "No EXIF metadata found" << endl;
        return;
    }

    // Decode EXIF into structured entries
    std::vector<std::vector<ExifEntry>> exif_entries;
    bool result = cv::decodeExif(raw_exif, exif_entries);
    if (!result)
    {
        cout << "Parsing EXIF raw data failed" << endl;
        return;
    }

    // Print original EXIF
    dumpExif(exif_entries, "ORIGINAL EXIF");

    // Encode image in-memory with original metadata
    std::vector<uchar> buf;
    cv::imencodeWithMetadata(".jpg", img, metadata_types, metadata, buf);

    // Re-encode EXIF from structured entries
    std::vector<uchar> raw_exif_data;
    bool res = cv::encodeExif(exif_entries, raw_exif_data);
    if (!res)
    {
        cout << "Failed to encode EXIF" << endl;
        return;
    }

    // Create new metadata vector with re-encoded EXIF
    std::vector<std::vector<uchar>> new_metadata;
    for (size_t i = 0; i < metadata_types.size(); i++)
    {
        if (metadata_types[i] == IMAGE_METADATA_EXIF)
            new_metadata.push_back(raw_exif_data); // Replace EXIF
        else
            new_metadata.push_back(metadata[i]);   // Keep other metadata unchanged
    }

    // Encode image again with new metadata
    cv::imencodeWithMetadata(".jpg", img, metadata_types, new_metadata, buf);

    // Decode the image from memory to verify metadata
    std::vector<int> metadata_types2;
    std::vector<std::vector<uchar>> metadata2;
    cv::Mat img2 = cv::imdecodeWithMetadata(buf, metadata_types2, metadata2);

    // Extract EXIF from reloaded image
    std::vector<uchar> raw_exif2 = getExifBlock(metadata_types2, metadata2);
    if (!raw_exif2.empty())
    {
        std::vector<std::vector<ExifEntry>> exif_entries2;
        if (cv::decodeExif(raw_exif2, exif_entries2))
        {
            // Print reloaded EXIF
            dumpExif(exif_entries2, "RELOADED EXIF");
        }
    }
}

int main(int argc, const char** argv)
{
    if (argc > 1)
        printExif(argv[1]);  // Process file from command line argument
    else
        cout << "usage : " << argv[0] << " filename";  // Print usage if no file given

    return 0;
}
